File: Binary Tree Level Order Traversal II.cpp
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {

        vector<vector<int>> res ;

        if (!root) return res ;

        queue<TreeNode*> q ;
        q.push(root) ;

        while ( !q.empty() ) {

            vector<int> dummy ;

            int size = q.size() ;

            while (size--) {
                TreeNode* curr = q.front() ; q.pop() ;
                dummy.push_back(curr->val) ;
                if (curr->left) q.push(curr->left) ;
                if (curr->right) q.push(curr->right) ;
            }

            res.push_back(dummy) ;

        }
        reverse(res.begin(), res.end());

        return res ;
    }
};

File: Binary Tree Level Order Traversal II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/*
Level Order Traversal
Time Complexity : O(n)
Space Complexity : O(n)
*/


class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {

        List<List<Integer>> result = new LinkedList<>() ;

        if (root == null) return result ;

        Deque<TreeNode> q = new ArrayDeque<>() ;
        q.offer(root) ;

        while (!q.isEmpty()) {
            int size = q.size() ;
            List<Integer> dummy = new LinkedList<Integer>() ;
            while (size-- > 0) {
                TreeNode curr = q.poll() ;
                dummy.add(curr.val) ;
                if (curr.left != null) q.offer(curr.left) ;
                if (curr.right != null) q.offer(curr.right) ;
            }
            result.add(0, dummy) ;
        }
        return result ;
    }
}